core/templates/components/forms/custom-forms-directives/thumbnail-uploader.component.ts

```diff
@@ -0,0 +1,265 @@
+  resampledFile: Blob;
+  newThumbnailDataUrl: string;
+  localStorageBgcolor: string;
+  imageUploadUrlTemplate: string;
+  editableThumbnailDataUrl = (
+    this.urlInterpolationService.getStaticImageUrl(
+      '/icons/story-image-icon.png'));
+  transformedData: string;
+  parsedResponse;
+
+  constructor(
+    private imageUploadHelperService: ImageUploadHelperService,
+    private alertsService: AlertsService,
+    private contextService: ContextService,
+    private imageLocalStorageService: ImageLocalStorageService,
+    private ngbModal: NgbModal,
+    private csrfTokenService: CsrfTokenService,
+    private urlInterpolationService: UrlInterpolationService
+  ) {}
+  placeholderImageDataUrl = (
+    this.urlInterpolationService.getStaticImageUrl(
+      '/icons/story-image-icon.png'));
+  thumbnailIsLoading = true;
+
+  ngOnInit(): void {
+    if (this.filename !== null &&
+        this.filename !== undefined &&
+        this.filename !== '') {
+      this.editableThumbnailDataUrl = (
+        this.imageUploadHelperService.getTrustedResourceUrlForThumbnailFilename(
+          this.filename,
+          this.contextService.getEntityType(),
+          this.contextService.getEntityId()));
+      this.uploadedImage = this.editableThumbnailDataUrl;
+    } else {
+      this.editableThumbnailDataUrl = this.placeholderImageDataUrl;
+    }
+  }
+
+  // 'ngOnChanges' is required here to update the thumbnail image
+  // everytime the thumbnail filename changes (eg. draft is discarded).
+  // The trusted resource url for the thumbnail should not be directly
+  // bound to ngSrc because it can cause an infinite digest error.
+  // This watcher is triggered only if the thumbnail filename of the
+  // model changes. It would change for the following operations:
+  // 1. Initial render of the page containing this directive.
+  // 2. When a thumbnail is uploaded.
+  // 3. When a saved draft is discarded.
+  ngOnChanges(changes: SimpleChanges): void {
+    if (
+      changes.filename &&
+      changes.filename.currentValue !== changes.filename.previousValue) {
+      const newValue = changes.filename.currentValue;
+      const previousValue = changes.filename.previousValue;
+      this.filenameChanges(newValue, previousValue);
+    }
+  }
+
+  filenameChanges(newFilename: string, prevFilename: string): void {
+    if (newFilename) {
+      this.editableThumbnailDataUrl = (
+        this.imageUploadHelperService
+          .getTrustedResourceUrlForThumbnailFilename(
+            newFilename,
+            this.contextService.getEntityType(),
+            this.contextService.getEntityId()));
+      this.uploadedImage = this.editableThumbnailDataUrl;
+    }
+    this.thumbnailIsLoading = true;
+  }
+
+  saveThumbnailBgColor(newBgColor: string): void {
+    if (newBgColor !== this.bgColor) {
+      this.updateBgColor.emit(newBgColor);
+    }
+  }
+
+  saveThumbnailImageData(imageURI: string, callback: {
+     (): void; (): void; }): void {
```

Comment:
what does `(): void; (): void;` type mean?

Reply:
fixed it